README.md
## Image carousel component for react-native
Based on https://github.com/nick/react-native-carousel .

### Installation
```bash
npm install https://github.com/ibc-corporate-travel/react-native-image-carousel
```

###Properties
#### Props were changed, actual readme coming soon.

```
hideIndicators={false} // Set to true to hide the indicators
indicatorStyle={color: '#000000', fontSize: 50} // Active indicator style
inactiveIndicatorStyle={color: '#999999', fontSize: 50} // Inactive indicator style
indicatorSpace={15} // space between each indicator
indicatorAtBottom={true} // Set to false to show the indicators at the top
indicatorOffset={250} // Indicator relative position from top or bottom
onPageChange={callback} // Called when the active page changes
inactiveIndicatorText= '•' // Inactive indicator content ( You can customize to use any Unicode character )
indicatorText= '•' // Active indicator content ( You can customize to use any Unicode character )
indicatorsPadding=3 // Amount of padding indicators on left and right sides;

animate={true} // Enable carousel autoplay
delay={1000} // Set Animation delay between slides
loop={true} // Allow infinite looped animation. Depends on Prop {...animate} set to true.

leftArr={'〈'}, // Left arr char. Can be text or React.Component.
rightArr={'〉'}, // Right arr char. Can be text or React.Component.
arrStyle={color: '#000000', fontSize: 30} // Styles for text(unicode?) arrows. It be applied only to text arrows!
arrowsTopOffset={50} // Arrows position from top

```


CHANGELOG.md
### 1.0.0-rc0
- props broken changes
- items declaration: used props instead of children
- fixed raced fetch
- used ListView for both ios and android
- added paddings for indicators
- added arrows
- project was rewritten with es2015
- massive refactoring

### 0.10.0
- Fix horizontal scroll in RN v0.28 (@sghiassy, @dmccer)

### 0.9.1
- onPageChange never called on indicator press on Android (@lideen)

### 0.9.0
- Fix carousel in iOS with react-native 0.28 (@francisco-sanchez-molina)

### 0.8.0
- Custom indicator content (@shidhincr)

### 0.7.0
- Call createClass on React instead of React Native (@scottluptowski)

### 0.6.1
- Hide inactive indicators (@scottluptowski)

### 0.6.0
- Compute width on the fly (@mikelambert)
- Android support (@mikelambert, @kushal)

### 0.5.0
- Fix for pressing indicator with custom width (@juankiz)
- Fix scrollWithoutAnimationTo deprecation warning (@juankiz)

### 0.4.6
- Fix depreciated parameter for scrollTo (@ronaldx)

### 0.4.5
- Fix scrolling to top when pressing iOS status bar (@juankiz)

### 0.4.4
- Added indicatorOffset property (@regeda)

### 0.4.3
- Fix for page loop animation

### 0.4.2
- Fix for Dimension import

### 0.4.1
- Remove duplicate width calculation when initialPage is defined (@ptmt)

### 0.4.0
- Add autoplay and looping options (@PraiseGeek)

### 0.3.5
- Add `indicatorSpace` prop (@alinz)

### 0.3.3
- Add `indicatorSize` prop (@daviferreira)

### 0.3.2
- Add `hideIndicators` prop (@daviferreira)

### 0.3.1
- Add `onPageChange` prop (@unknownexception)

### 0.3.0
- Bump react-native dependency

### 0.2.0
- Add `indicatorAtBottom` prop (@imaffett)
- Added new styles for the  indicator positions (@imaffett)
- Clicking the indicators will switch the card/page it is on. (@imaffett)
- Set the background of the indicator to transparent. (@imaffett)
- Set bounces=false on the ScrollView used for the Carousel (@imaffett)

### 0.1.0
- Initial component


CarouselPager.js
import React, {Component} from 'react';
import {ListView, TouchableOpacity, Image, ActivityIndicator, StyleSheet} from 'react-native';
import Dimensions from 'Dimensions';


const windowWidth = Dimensions.get('window').width;

const styles = StyleSheet.create({
  image: {
    height: 200,
    width: windowWidth
  },
  imageContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    width: windowWidth
  }
});

const ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});

class ImageCell extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isImageLoaded: false
    };
  }

  render() {
    const {uri, onPress} = this.props;
    const {isImageLoaded} = this.state;

    return (
      <TouchableOpacity style={styles.imageContainer} onPress={onPress}>
        <ActivityIndicator
          animating = {!isImageLoaded}
          style={{
            alignItems: 'center',
            justifyContent: 'center',
            position: 'absolute',
            left: 0,
            top: 0,
            right: 0,
            bottom: 0
          }}
        />
        <Image
          onLoadEnd   = {() => this.setState({isImageLoaded: true})}
          source      = {{uri}}
          style       = {styles.image}
          resizeMode  = "cover"
        />
      </TouchableOpacity>
    );
  }
};

var CarouselPager = React.createClass({

  scrollToPage(page, animated) {
    if (typeof animated === 'undefined') {
      animated = true;
    }
    this.refs.cont.scrollTo({x: page * this.props.width, y: 0, animated: animated});
  },

  _onMomentumScrollEnd(e) {
    var activePage = e.nativeEvent.contentOffset.x / this.props.width;
    this.props.onEnd(activePage);
  },

  render() {
    const {images, onPress} = this.props;

    return (
      <ListView
        ref                               = "cont"
        dataSource                        = {ds.cloneWithRows(images.map((uri, i) => ({uri, i})))}
        renderRow                         = {({uri, i}) => <ImageCell uri={uri} onPress={() => onPress(i)}/>}
        initialListSize                   = {1}
        pageSize                          = {1}
        automaticallyAdjustContentInsets  = {false}
        showsHorizontalScrollIndicator    = {false}
        bounces                           = {false}
        scrollsToTop                      = {false}
        contentContainerStyle             = {this.props.contentContainerStyle}
        onScrollBeginDrag                 = {this.props.onBegin}
        onMomentumScrollEnd               = {this._onMomentumScrollEnd}
        horizontal
        pagingEnabled
      />
    );
  },
});

module.exports = CarouselPager;


src/CarouselPager/index.js
import React, {Component} from 'react';
import {ListView} from 'react-native';
import CarouselItem from '../CarouselItem';

const ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});

class CarouselPager extends Component {
  constructor(props) {
    super(props);
    this._onMomentumScrollEnd = this._onMomentumScrollEnd.bind(this);
  }

  scrollToPage(page, animated) {
    if (typeof animated === 'undefined') {
      animated = true;
    }
    this.refs.cont.scrollTo({x: page * this.props.width, y: 0, animated: animated});
  }

  _onMomentumScrollEnd(e) {
    const {width, onEnd} = this.props;
    const activePage = e.nativeEvent.contentOffset.x / width;
    onEnd(activePage);
  }

  render() {
    const {
      images,
      onPress,
      contentContainerStyle,
      onBegin
  } = this.props;

    return (
      <ListView
        ref                               = "cont"
        dataSource                        = {ds.cloneWithRows(images.map((uri, i) => ({uri, i})))}
        renderRow                         = {({uri, i}) => <CarouselItem uri={uri} onPress={() => onPress(i)}/>}
        initialListSize                   = {1}
        pageSize                          = {1}
        automaticallyAdjustContentInsets  = {false}
        showsHorizontalScrollIndicator    = {false}
        bounces                           = {false}
        scrollsToTop                      = {false}
        contentContainerStyle             = {contentContainerStyle}
        onScrollBeginDrag                 = {onBegin}
        onMomentumScrollEnd               = {this._onMomentumScrollEnd}
        horizontal
        pagingEnabled
      />
    );
  }
};

export default CarouselPager;


Carousel.js
'use strict';

var React = require('react');
var {
  Dimensions,
  Text,
  View,
  TouchableOpacity
} = require('react-native');

var TimerMixin = require('react-timer-mixin');
var CarouselPager = require('./CarouselPager');
var styles = require('./styles');

var Carousel = React.createClass({
  mixins: [TimerMixin],

  getDefaultProps() {
    return {
      hideIndicators: false,
      indicatorStyle: styles.defaultIndicatorStyle,
      inactiveIndicatorStyle: styles.defaultInactiveIndicatorStyle,
      indicatorAtBottom: true,
      indicatorOffset: 250,
      indicatorText: '•',
      inactiveIndicatorText: '•',
      indicatorsPadding: 3,
      width: null,
      initialPage: 0,
      indicatorSpace: 25,
      animate: true,
      delay: 1000,
      loop: true,
      rightArr: '〉',
      leftArr: '〈',
      arrStyle: styles.defaultArr,
      arrowsTopOffset: 50,
    };
  },

  getInitialState() {
    return {
      activePage: this.props.initialPage > 0 ? this.props.initialPage : 0,
    };
  },

  getWidth() {
    if (this.props.width !== null) {
      return this.props.width;
    } else {
      return Dimensions.get('window').width;
    }
  },

  componentDidMount() {
    if (this.props.initialPage > 0) {
      this.refs.pager.scrollToPage(this.props.initialPage, false);
    }

    if (this.props.animate && this.props.images) {
      this._setUpTimer();
    }
  },

  scrollTo(activePage) {
    this.setState({activePage});
    this.refs.pager.scrollToPage(activePage);
  },

  wrapArrow(arrow) {
    var {arrStyle} = this.props;

    if (typeof arrow === 'string') {
      //text or unicode char
      return <Text style={[styles.defaultArr, arrStyle]}>
        {arrow}
      </Text>;
    } else {
      //component
      return arrow;
    }
  },

  calcIndicatorsPosition(indicatorContWidth) {
    const isWidthContainIndicators = this.getWidth() > indicatorContWidth;
    if (isWidthContainIndicators) {
      return (this.getWidth() - indicatorContWidth) / 2;
    }

    const {indicatorSpace} = this.props;
    const {activePage} = this.state;
    const spaceToActiveIndicator = activePage * indicatorSpace;

    const diff = spaceToActiveIndicator - this.getWidth();
    if (diff >= 0) {
      return -(diff + indicatorSpace);
    } else {
      return 0;
    }
  },

  addPaddingsToIndicators(indicators, indicatorContWidth) {
    const isWidthContainIndicators = this.getWidth() > indicatorContWidth;
    if (isWidthContainIndicators) return indicators;

    const {
      images,
      indicatorsPadding,
      inactiveIndicatorStyle,
      inactiveIndicatorText
    } = this.props;

    for (let i = 1, l = indicatorsPadding; i <= l; i++) {
      indicators.unshift(
        <Text
          style   = {inactiveIndicatorStyle}
          key     = {`paddingLeft_${i}`}
          onPress = {() => this.scrollTo(images.length - i)}
        >
          {inactiveIndicatorText}
        </Text>
      );

      indicators.push(
        <Text
          style   = {inactiveIndicatorStyle}
          key     = {`paddingRight_${i}`}
          onPress = {() => this.scrollTo(i - 1)}
        >
          {inactiveIndicatorText}
        </Text>
      )
    }

    return indicators;
  },

  renderPageIndicator() {
    const {hideIndicators} = this.props;
    if (hideIndicators) return null;

    const {images} = this.props;
    const {activePage} = this.state;

    let indicators = [];
    for (var i = 0, l = images.length; i < l; i++) {
      const isActive  = i === activePage;

      const {indicatorStyle, inactiveIndicatorStyle} = this.props;
      const style = isActive ? indicatorStyle : inactiveIndicatorStyle;

      const {indicatorText, inactiveIndicatorText} = this.props;
      const text = isActive ? indicatorText : inactiveIndicatorText;

      indicators.push(
        <Text
          style   = {style}
          key     = {i}
          onPress = {this.scrollTo.bind(this, i)}
        >
          {text}
        </Text>
      );
    }

    if (indicators.length === 1) return null;

    const {indicatorSpace} = this.props;
    const indicatorContWidth = images.length * indicatorSpace;

    const indicatorsContStyle = {
      width: indicatorContWidth,
      left: this.calcIndicatorsPosition(indicatorContWidth)
    };

    const {indicatorAtBottom, indicatorOffset} = this.props;
    if (indicatorAtBottom) {
      indicatorsContStyle.bottom = indicatorOffset;
    } else {
      indicatorsContStyle.top = indicatorOffset;
    }

    indicators = this.addPaddingsToIndicators(indicators, indicatorContWidth);

    return <View style={[styles.pageIndicator, indicatorsContStyle]}>
      {indicators}
    </View>;
  },

  renderLeftArr() {
    var {
      leftArr,
      arrowsTopOffset,
      images,
      loop
    } = this.props;

    var {activePage} = this.state;

    var prev = activePage - 1;

    var onPress;
    if (prev >= 0) {
      onPress = () => this.scrollTo(prev);
    } else {
      onPress = loop
        ? () => this.scrollTo(images.length - 1)
        : () => null;
    }


    return (
      <View style={[styles.leftArrCont, {top: arrowsTopOffset}]}>
        <TouchableOpacity onPress={onPress}>
          {this.wrapArrow(leftArr)}
        </TouchableOpacity>
      </View>
    );
  },

  renderRightArr() {
    var {
      rightArr,
      arrowsTopOffset,
      images,
      loop
    } = this.props;

    var {activePage} = this.state;

    var next = activePage + 1;

    var onPress;
    if (next < images.length) {
      onPress = () => this.scrollTo(next);
    } else {
      onPress = loop
        ? () => this.scrollTo(0)
        : () => null;
    }

    return (
      <View style={[styles.rightArrCont, {top: arrowsTopOffset}]}>
        <TouchableOpacity onPress={onPress}>
          {this.wrapArrow(rightArr)}
        </TouchableOpacity>
      </View>
    );
  },

  _setUpTimer() {
    if (this.props.images.length > 1) {
      this.clearTimeout(this.timer);
      this.timer = this.setTimeout(this._animateNextPage, this.props.delay);
    }
  },

  _animateNextPage() {
    var activePage = 0;
    if (this.state.activePage < this.props.images.length - 1) {
      activePage = this.state.activePage + 1;
    } else if (!this.props.loop) {
      return;
    }

    this.scrollTo(activePage);
    this._setUpTimer();
  },

  _onAnimationBegin() {
    this.clearTimeout(this.timer);
  },

  _onAnimationEnd(activePage) {
    this.setState({activePage});
    if (this.props.onPageChange) {
      this.props.onPageChange(activePage);
    }
  },

  render() {
    return (
      <View style={{ flex: 1 }}>
        <CarouselPager
          onPress = {this.props.onPress}
          images = {this.props.images}
          ref="pager"
          width={this.getWidth()}
          contentContainerStyle={styles.container}
          onBegin={this._onAnimationBeginPage}
          onEnd={this._onAnimationEnd}
        />
        {this.renderPageIndicator()}
        {this.renderLeftArr()}
        {this.renderRightArr()}
      </View>
    );
  },
});

module.exports = Carousel;
